### PR TITLE
test FIPS mode on centos8

### PR DESCRIPTION
**Pull request by reaperhulk**
replaces #5024 

This has been accreting across multiple PRs for a long time so we should also think about whether the general approach is actually what we want here.

### CODE REVIEW

File: tests/hazmat/primitives/test_dh.py [hunk 1]
```diff
@@ -29,6 +29,30 @@ def _skip_dhx_unsupported(backend, is_dhx):
         )
 
 
+FFDH3072_P = int_from_bytes(
```
Comment (reaperhulk) [inline tests/hazmat/primitives/test_dh.py:32]: (Not sure why this was necessary as a fixture -- needs investigation.

File: tests/hazmat/primitives/test_pkcs12.py [hunk 2]
```diff
@@ -55,17 +55,20 @@ def test_load_pkcs12_ec_keys(self, filename, password, backend):
         self._test_load_pkcs12_ec_keys(filename, password, backend)
 
     @pytest.mark.parametrize(
-        ("filename", "password"),
+        ("filename", "password", "fips"),
         [
-            ("cert-rc2-key-3des.p12", b"cryptography"),
-            ("no-password.p12", None),
+            ("cert-rc2-key-3des.p12", b"cryptography", False),
+            ("no-password.p12", None, False), # TODO: this should be false, but what happens?
```
Comment (reaperhulk) [inline tests/hazmat/primitives/test_pkcs12.py:61]: Remove this outdated comment

File: tests/hazmat/primitives/test_rsa.py [hunk 3]
```diff
@@ -152,6 +153,11 @@ class TestRSA(object):
         )
     )
     def test_generate_rsa_keys(self, backend, public_exponent, key_size):
+        if backend._fips_enabled:
+            if key_size < backend._rsa_min_key_size:
+                pytest.skip("Small key size blocked by crypto policy")
```
Comment (reaperhulk) [inline tests/hazmat/primitives/test_rsa.py:158]: We should just say FIPS here so it's not confusing

File: tests/hazmat/primitives/test_rsa.py [hunk 4]
```diff
@@ -152,6 +153,11 @@ class TestRSA(object):
...
+            if public_exponent < backend._rsa_min_public_exponent:
+                pytest.skip("Small exponent blocked by crypto policy")
```
Comment (reaperhulk) [inline tests/hazmat/primitives/test_rsa.py:160]: Same

File: docs/hazmat/backends/openssl.rst [hunk 5]
```diff
@@ -125,6 +125,62 @@ seeded from the same pool as ``/dev/random``.
 +------------------------------------------+------------------------------+
 
 
+FIPS mode
+---------
+
+Federal Information Processing Standards is a public standard for use by
+US American government agencies. OpenSSL in FIPS mode prohibits the use of
+non-approved algorithms and parameters.
```
Comment (alex) [inline docs/hazmat/backends/openssl.rst:133]: This should a) indicate it means FIPS-140, there's a lot of other FIPS we don't care about.

Also I want it to say that we think FIPS is garbage, and you shouldn't use it unless you absolutely have to.

File: docs/hazmat/backends/openssl.rst [hunk 6]
```diff
@@ -125,6 +125,62 @@ seeded from the same pool as ``/dev/random``.
...
+
+Key restrictions
+~~~~~~~~~~~~~~~~
+
+* _`Diffie-Hellman key exchange` requires key size equal or greater than
```
Comment (alex) [inline docs/hazmat/backends/openssl.rst:138]:
```suggestion
* `Diffie-Hellman key exchange`_ requires key size equal or greater than
```

you had the wrong link syntax

Comment (alex) [reply docs/hazmat/backends/openssl.rst:138]: Though I have no idea where this is supposed to link?

File: docs/hazmat/backends/openssl.rst [hunk 7]
```diff
@@ -125,6 +125,62 @@ seeded from the same pool as ``/dev/random``.
...
+  2048 bits.
+* _`DSA` requires key size equal or greater than 2048 bits.
+* _`RSA`  requires key size equal or greater than 2048 bits and the
+  public exponent must be 65537.
+
+Prohibited message digests
```
Comment (alex) [inline docs/hazmat/backends/openssl.rst:144]: Wouldn't this be easier if we listed _approved_ algorithms, not banned ones?

FIPS is expressed as an approval list, so this is honestly confusing.

File: docs/hazmat/backends/openssl.rst [hunk 8]
```diff
@@ -125,6 +125,62 @@ seeded from the same pool as ``/dev/random``.
...
+* _`Ed25519 signing`
+* _`X448 key exchange`
+* _`X25519 key exchange`
+
+Blocked features
+~~~~~~~~~~~~~~~~
+
+* Encryption with
+  :attr:`~cryptography.hazmat.primitives.serialization.PrivateFormat.TraditionalOpenSSL`
+  format
```
Comment (alex) [inline docs/hazmat/backends/openssl.rst:180]: This is confusing to me.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 9]
```diff
@@ -119,21 +120,50 @@ class Backend(object):
     """
     name = "openssl"
 
+    # FIPS doesn't allow several algorithms or weak key sizes, but the
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:123]: same concern, I strongly feel we should express this as a list of approved keys.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 10]
```diff
@@ -119,21 +120,50 @@ class Backend(object):
...
+    _rsa_min_public_exponent = 65537
+    _dsa_min_modulus = 1 << 2048
+    _dh_min_key_size = 2048
+    _dh_min_modulus = 1 << _dh_min_key_size
+
     def __init__(self):
         self._binding = binding.Binding()
         self._ffi = self._binding.ffi
         self._lib = self._binding.lib
+        self._fips_enabled = self._is_fips_enabled()
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:141]: This means if you try to enable FIPS later, won't something confusing happen?

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 11]
```diff
@@ -119,21 +120,50 @@ class Backend(object):
...
 
         self._cipher_registry = {}
         self._register_default_ciphers()
-        self.activate_osrandom_engine()
+        if self._fips_enabled:
+            warnings.warn(
+                "OpenSSL FIPS mode is enabled. Can't enable DRBG fork safety.",
+                UserWarning
+            )
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:149]: a) This is wrong, we should only do this warning in cases where we're actually going to be use that engine (e.g. openssl<1.1.1d)
b) Should we even allow FIPS on OpenSSL's that have a broken internal CSPRNG? That's literally enabling a vulnerability.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 12]
```diff
@@ -119,21 +120,50 @@ class Backend(object):
...
     def openssl_assert(self, ok):
         return binding._openssl_assert(self._lib, ok)
 
+    def _is_fips_enabled(self):
+        fips_mode = getattr(self._lib, "FIPS_mode", lambda: 0)
+        mode = fips_mode()
+        if mode == 0:
+            # OpenSSL without FIPS pushes an error on the error stack
+            self._lib.ERR_clear_error()
+        return mode
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:165]:
```suggestion
        return bool(mode)
```

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 13]
```diff
@@ -1937,6 +1987,13 @@ def generate_dh_parameters(self, generator, key_size):
         if key_size < 512:
             raise ValueError("DH key_size must be at least 512 bits")
 
+        if False and self._fips_enabled and key_size < self._dh_min_key_size:
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:1990]: TODO

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 14]
```diff
@@ -2170,6 +2227,8 @@ def x25519_generate_key(self):
         return _X25519PrivateKey(self, evp_pkey)
 
     def x25519_supported(self):
+        if self._fips_enabled:
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:2230]: These are all things that are going to need to be updated once FIPS-186-5 is standardized. I'm extremely frustrated that this is require to require so much ongoing maintenance.

File: tests/hazmat/backends/test_openssl.py [hunk 15]
```diff
@@ -171,6 +171,9 @@ def test_bn_to_int(self):
 @pytest.mark.skipif(
     not backend._lib.CRYPTOGRAPHY_NEEDS_OSRANDOM_ENGINE,
     reason="Requires OpenSSL with ENGINE support and OpenSSL < 1.1.1d")
+@pytest.mark.skipif(
```
Comment (alex) [inline tests/hazmat/backends/test_openssl.py:174]: Should this use `skip_fips`?

File: tests/conftest.py [hunk 16]
```diff
@@ -33,6 +36,15 @@ def pytest_generate_tests(metafunc):
         metafunc.parametrize("wycheproof", testcases)
 
 
+def pytest_runtest_setup(item):
+    if openssl_backend._fips_enabled:
+        for marker in item.iter_markers(name="skip_fips"):
+            if "reason" in marker.kwargs:
```
Comment (alex) [inline tests/conftest.py:42]: Any reason not to require a reason?

File: tests/hazmat/backends/test_openssl_memleak.py [hunk 17]
```diff
@@ -298,6 +300,7 @@ def func():
             ec.derive_private_key(1, ec.SECP256R1(), backend)
         """))
 
+    @pytest.mark.skip_fips
```
Comment (alex) [inline tests/hazmat/backends/test_openssl_memleak.py:303]: This whole class is skipped?

File: tests/hazmat/primitives/fixtures_dh.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
+# This file is dual licensed under the terms of the Apache License, Version
```
Comment (alex) [inline tests/hazmat/primitives/fixtures_dh.py:1]: Can you put this refactor in it's own PR?

File: tests/hazmat/primitives/test_dh.py [hunk 19]
```diff
@@ -16,10 +16,16 @@
 from cryptography.hazmat.primitives.asymmetric import dh
 from cryptography.utils import int_from_bytes
 
+from .fixtures_dh import FFDH3072_P
 from ...doubles import DummyKeySerializationEncryption
 from ...utils import load_nist_vectors, load_vectors_from_file
 
 
+@pytest.fixture()
```
Comment (alex) [inline tests/hazmat/primitives/test_dh.py:24]: I'm not a big fan of this fixtures style. Is there a reason not to just call a function in the test?

File: tests/hazmat/primitives/test_pkcs12.py [hunk 20]
```diff
@@ -55,17 +55,20 @@ def test_load_pkcs12_ec_keys(self, filename, password, backend):
         self._test_load_pkcs12_ec_keys(filename, password, backend)
 
     @pytest.mark.parametrize(
-        ("filename", "password"),
+        ("filename", "password", "fips"),
```
Comment (alex) [inline tests/hazmat/primitives/test_pkcs12.py:58]: I do not understand what FIPS means in this conext. Please more it more clear.

File: tests/hazmat/primitives/utils.py [hunk 21]
```diff
@@ -470,3 +479,13 @@ def _check_dsa_private_numbers(skey):
     pkey = skey.public_numbers
     params = pkey.parameter_numbers
     assert pow(params.g, skey.x, params.p) == pkey.y
+
+
+def skip_fips_traditional_openssl(backend, fmt):
+    if (
+            fmt is serialization.PrivateFormat.TraditionalOpenSSL and
```
Comment (alex) [inline tests/hazmat/primitives/utils.py:486]: over indented.

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 11 above)
Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/backend.py:149]: We absolutely must. There exist **no FIPS** openssl releases/validated distributions with the reseeding PRNG

Comment (alex) [reply src/cryptography/hazmat/backends/openssl/backend.py:149]: So there is no way to safely use a FIPS OpenSSL if your program forks?

Remind me why we're putting any work at all into this? I've worked for the US Government before, and I gotta say, if you need FIPS, maybe you should fucking do the work yourself instead of making volunteer OSS maintainers deal with NIST's braindamage.

File: docs/hazmat/backends/openssl.rst (hunk 5 above)
Comment (reaperhulk) [reply docs/hazmat/backends/openssl.rst:133]: I'm tempted to say we shouldn't have a FIPS section at all. If you need FIPS you have a special set of nonsense requirements and what version you're bound by and what matters is ultimately between you and the auditor. I'd rather say nothing right now and maybe later have some limited statements about how we handle FIPS. What do you think?

Comment (alex) [reply docs/hazmat/backends/openssl.rst:133]: That's fine with me, although just to be clear, most federal government contexts in which FIPS wil be (nominally) required don't actually have an auditor. They've got a person responsible for rubber stamping your paperwork, but no one will ever log into a computer or read a line of course code.

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 9 above)
Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/backend.py:123]: (Switched to allow list)

File: tests/hazmat/primitives/test_dsa.py [hunk 22]
```diff
@@ -40,6 +41,13 @@ def _skip_if_dsa_not_supported(backend, algorithm, p, q, g):
         )
 
 
+def _min_key_size(backend):
+    if backend._fips_enabled:
+        return backend._dh_min_key_size
+    else:
+        return 1024
```
Comment (alex) [inline tests/hazmat/primitives/test_dsa.py:48]: Is there a reason not to always just use 2048 here?

File: tests/hazmat/primitives/test_ec.py [hunk 23]
```diff
@@ -45,17 +46,18 @@ def _skip_ecdsa_vector(backend, curve_type, hash_type):
         curve_type()
     ):
         pytest.skip(
-            "ECDSA not supported with this hash {} and curve {}".format(
-                hash_type().name, curve_type().name
+            "ECDSA not supported with this hash {} and curve {}. "
+            "FIPS mode: {}".format(
```
Comment (alex) [inline tests/hazmat/primitives/test_ec.py:50]: It's extremely annoying to need to scatter this everywhere. Can we add it to the backend repr and just include the backend in the skip or something?

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 24]
```diff
@@ -119,21 +120,59 @@ class Backend(object):
...
+        hashes.SHA1, hashes.SHA224, hashes.SHA256, hashes.SHA384,
+        hashes.SHA512, hashes.SHA512_224, hashes.SHA512_256, hashes.SHA3_224,
+        hashes.SHA3_256, hashes.SHA3_384, hashes.SHA3_512, hashes.SHAKE128,
+        hashes.SHAKE256,
+    )
+    _rsa_min_key_size = 2048
+    _rsa_min_public_exponent = 65537
+    _dsa_min_modulus = 1 << 2048
+    _dh_min_key_size = 2048
+    _dh_min_modulus = 1 << _dh_min_key_size
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:144]: Should these all mention FIPS?

File: tests/hazmat/primitives/test_dsa.py (hunk 22 above)
Comment (reaperhulk) [reply tests/hazmat/primitives/test_dsa.py:48]: Generating a 2048-bit DH key is annoyingly slow, but avoiding the branch is probably worth it.

File: tests/hazmat/primitives/test_ec.py (hunk 23 above)
Comment (reaperhulk) [reply tests/hazmat/primitives/test_ec.py:50]: Our backends don't have a repr! (This is an issue I noticed earlier today, lol)

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 10 above)
Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/backend.py:150]: For now, yes. It's easier to require that it be in FIPS mode at startup (so we don't try to add the engine at all), so let's just make that a requirement.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 25]
```diff
@@ -2398,6 +2467,8 @@ def load_key_and_certificates_from_pkcs12(self, data, password):
         return (key, cert, additional_certificates)
 
     def poly1305_supported(self):
+        if self._fips_enabled:
+            return 0
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py:2471]:
```suggestion
            return False
```

File: tests/hazmat/primitives/test_dsa.py [hunk 26]
```diff
@@ -49,7 +50,8 @@ def test_skip_if_dsa_not_supported(backend):
 @pytest.mark.requires_backend_interface(interface=DSABackend)
 class TestDSA(object):
     def test_generate_dsa_parameters(self, backend):
-        parameters = dsa.generate_parameters(1024, backend)
+        key_size = 2048
+        parameters = dsa.generate_parameters(key_size, backend)
```
Comment (alex) [inline tests/hazmat/primitives/test_dsa.py:54]:
```suggestion
        parameters = dsa.generate_parameters(2048, backend)
```

File: tests/hazmat/primitives/test_dsa.py [hunk 27]
```diff
@@ -91,7 +98,8 @@ def test_generate_dsa_keys(self, vector, backend):
         )
 
     def test_generate_dsa_private_key_and_parameters(self, backend):
-        skey = dsa.generate_private_key(1024, backend)
+        key_size = 2048
+        skey = dsa.generate_private_key(key_size, backend)
```
Comment (alex) [inline tests/hazmat/primitives/test_dsa.py:102]:
```suggestion
        skey = dsa.generate_private_key(2048, backend)
```

File: tests/hazmat/primitives/utils.py [hunk 28]
```diff
@@ -80,6 +81,14 @@ def test_aead(self, backend, params):
 
 
 def aead_test(backend, cipher_factory, mode_factory, params):
+    if (
```
Comment (alex) [inline tests/hazmat/primitives/utils.py:84]: Can you add a comment indicating that the check is for 24, not 12, because hex?
